src/provider/provider_level_zero.c

```diff
@@ -0,0 +1,240 @@
+    ze_result_t (*zeMemFree)(ze_context_handle_t, void *);
+} ze_ops_t;
+
+static ze_ops_t g_ze_ops;
+static UTIL_ONCE_FLAG ze_is_initialized = UTIL_ONCE_FLAG_INIT;
+static bool Init_ze_global_state_failed;
+
+static void init_ze_global_state(void) {
+    // check if Level Zero shared library is already loaded
+    *(void **)&g_ze_ops.zeMemAllocDevice =
+        dlsym(RTLD_DEFAULT, "zeMemAllocDevice");
+    *(void **)&g_ze_ops.zeMemFree = dlsym(RTLD_DEFAULT, "zeMemFree");
+
+    if (!g_ze_ops.zeMemAllocDevice || !g_ze_ops.zeMemFree) {
+        fprintf(stderr, "Required Level Zero symbols not found.\n");
+        Init_ze_global_state_failed = true;
+    }
+}
+
+enum umf_result_t ze_memory_provider_initialize(void *params, void **provider) {
+    if (provider == NULL || params == NULL) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    level_zero_memory_provider_params_t *ze_params =
+        (level_zero_memory_provider_params_t *)params;
+
+    util_init_once(&ze_is_initialized, init_ze_global_state);
+    if (Init_ze_global_state_failed) {
+        fprintf(stderr, "Loading Level Zero symbols failed\n");
+        return UMF_RESULT_ERROR_UNKNOWN;
+    }
+
+    if (ze_params->memory_type != UMF_MEMORY_TYPE_DEVICE) {
+        // TODO support other memory types
+        return UMF_RESULT_ERROR_NOT_SUPPORTED;
+    }
+
+    ze_memory_provider_t *ze_provider =
+        umf_ba_global_alloc(sizeof(ze_memory_provider_t));
+    if (!ze_provider) {
+        return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+    }
+
+    ze_provider->context = ze_params->level_zero_context_handle;
+    ze_provider->device = ze_params->level_zero_device_handle;
+    ze_provider->memory_type = ze_params->memory_type;
+    ze_provider->host_mem_alloc_desc = (ze_host_mem_alloc_desc_t){
+        ZE_STRUCTURE_TYPE_HOST_MEM_ALLOC_DESC,
+        NULL, // pointer to an extension-specific structure
+        ze_params->level_zero_host_mem_alloc_flags};
+    ze_provider->device_mem_alloc_desc = (ze_device_mem_alloc_desc_t){
+        ZE_STRUCTURE_TYPE_DEVICE_MEM_ALLOC_DESC,
+        NULL, // pointer to an extension-specific structure
+        ze_params->level_zero_device_mem_alloc_flags,
+        ze_params->level_zero_device_local_mem_ordinal};
+
+    *provider = ze_provider;
+
+    return UMF_RESULT_SUCCESS;
+}
+
+void ze_memory_provider_finalize(void *provider) {
+    assert(provider);
+
+    util_init_once(&ze_is_initialized, init_ze_global_state);
+    umf_ba_global_free(provider, sizeof(ze_memory_provider_t));
+}
+
+static enum umf_result_t ze_memory_provider_alloc(void *provider, size_t size,
+                                                  size_t alignment,
+                                                  void **resultPtr) {
+    assert(provider);
+    assert(resultPtr);
+
+    ze_memory_provider_t *ze_provider = (ze_memory_provider_t *)provider;
+    ze_result_t ze_result = g_ze_ops.zeMemAllocDevice(
```

Comment:
Based on `ze_provider->memory_type` we should use `zeMemAllocDevice`, `zeMemAllocHost` or `zeMemAllocShared`.

Reply:
done